src/parser.ts

```diff
@@ -0,0 +1,76 @@
+import $RefParser from '@apidevtools/json-schema-ref-parser';
+import { JSONPath } from 'jsonpath-plus';
+import { merge } from 'lodash';
+
+//import { Schema } from '@asyncapi/parser';
+import { JSONSchema7 } from 'json-schema';
+
+class ExternalComponents {
+    ref;
+    resolvedJSON;
+    constructor(ref: string, resolvedJSON: string) {
+        this.ref = ref;
+        this.resolvedJSON = resolvedJSON;
+    }
+
+    getKey() {
+        const keys = this.ref.split('/');
+        return keys[keys.length - 1]
+    }
+
+    getValue() {
+        return this.resolvedJSON;
+    }
+}
+
+/**
+ * resolves external references and updates $refs 
+ * @param {Object[]} JSONSchema 
+ */
+export async function parse(JSONSchema: JSONSchema7) {
+    const $ref = await $RefParser.resolve(JSONSchema);
+    const refs = crawlChanelPropertiesForRefs(JSONSchema);
+    for (let ref of refs) {
+        if (isExternalReference(ref)) {
+            const componentObject = await resolveExternalRefs(JSONSchema, $ref);
+            if (JSONSchema.components) {
+                merge(JSONSchema.components, componentObject);
+            } else {
+                JSONSchema.components = componentObject
+            }
+        }
+    }
+}
+
+function crawlChanelPropertiesForRefs(JSONSchema: JSONSchema7) {
+    return JSONPath({ json: JSONSchema, path: `$.channels.*.*.message['$ref']` });
+}
+
+/**
+ * This function checks for external reference. 
+ * @param {string} ref 
+ * @returns {boolean}
+ */
+function isExternalReference(ref: string) {
+    return !ref.startsWith('#')
+}
+
+/**
+ * 
+ * @param {Object[]} parsedJSON 
+ * @param {$RefParser} $refs 
+ * @returns {ExternalComponents}
+ */
+async function resolveExternalRefs(parsedJSON: JSONSchema7, $refs: JSONSchema7) {
+    const componentObj = { messages: {} };
```

Comment:
Here you have to define the type for the component object, This will basically look like AsyncAPI's component object, you can refer our documentation https://github.com/asyncapi/spec/blob/v2.4.0/spec/asyncapi.md#componentsObject to know more about it. 

Something like this would work

```ts
interface IComponent {
    messages: Record<string, any>
}

```